### PR TITLE
[MPS] topk+kth value metal kernels

### LINKED ISSUE #154890: MPS topk failure for 5D tensor or above
### 🐛 Describe the bug

Similar to #154881 

```python
import torch

x = torch.ones([5, 4, 3, 2, 1], device="mps")
torch.ops.aten.topk(x, k=5, dim=0)
```

Error message:
```
/AppleInternal/Library/BuildRoots/01adf19d-fba1-11ef-a947-f2a857e00a32/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArraySort.mm:208: failed assertion `(null)" Axis = 4. This class only supports axis = 0, 1, 2, 3
'
zsh: abort
```

### Versions

torch==2.7.0

### PR DESCRIPTION
**Pull request by Isalia20**
topk + kth value metal kernels using the sort kernel helpers. Fixes #154890

**float32**

| Shape | dim | layout | topk k=8 | topk k=256 | kthvalue k=N/2 |
|---|---|---|---|---|---|
| `[8, 256]` | -1 | contiguous | 0.77x | 1.79x | 1.73x |
| `[8, 1024]` | -1 | contiguous | 1.99x | 2.10x | 1.99x |
| `[8, 4096]` | -1 | contiguous | 1.41x | 1.47x | 1.39x |
| `[8, 16384]` | -1 | contiguous | 1.90x | 1.96x | 1.93x |
| `[8, 65536]` | -1 | contiguous | 1.14x | 1.14x | 1.14x |
| `[128, 256]` | -1 | contiguous | 7.76x | 7.64x | 7.71x |
| `[128, 1024]` | -1 | contiguous | 4.22x | 4.23x | 4.17x |
| `[128, 4096]` | -1 | contiguous | 1.36x | 1.36x | 1.35x |
| `[128, 16384]` | -1 | contiguous | 1.16x | 1.17x | 1.18x |
| `[128, 65536]` | -1 | contiguous | 1.36x | 1.44x | 1.43x |
| `[1024, 256]` | -1 | contiguous | 10.97x | 10.79x | 10.74x |
| `[1024, 1024]` | -1 | contiguous | 3.95x | 3.97x | 3.94x |
| `[1024, 4096]` | -1 | contiguous | 1.27x | 1.28x | 1.27x |
| `[1024, 16384]` | -1 | contiguous | 1.11x | 1.11x | 1.10x |
| `[1024, 65536]` | -1 | contiguous | 1.89x | 1.76x | 1.88x |
| `[8, 256]` | -1 | non-contig | 1.87x | 1.75x | 1.78x |
| `[8, 1024]` | -1 | non-contig | 2.05x | 2.01x | 2.07x |
| `[8, 4096]` | -1 | non-contig | 1.44x | 1.46x | 1.42x |
| `[8, 16384]` | -1 | non-contig | 1.98x | 1.96x | 1.84x |
| `[8, 65536]` | -1 | non-contig | 1.31x | 1.29x | 1.30x |
| `[128, 256]` | -1 | non-contig | 7.87x | 7.23x | 7.58x |
| `[128, 1024]` | -1 | non-contig | 4.17x | 4.20x | 4.25x |
| `[128, 4096]` | -1 | non-contig | 1.34x | 1.33x | 1.35x |
| `[128, 16384]` | -1 | non-contig | 1.30x | 1.31x | 1.31x |
| `[128, 65536]` | -1 | non-contig | 1.97x | 1.88x | 1.92x |
| `[1024, 256]` | -1 | non-contig | 10.92x | 11.22x | 10.80x |
| `[1024, 1024]` | -1 | non-contig | 4.03x | 3.98x | 3.93x |
| `[1024, 4096]` | -1 | non-contig | 1.30x | 1.31x | 1.30x |
| `[1024, 16384]` | -1 | non-contig | 1.28x | 1.28x | 1.28x |
| `[1024, 65536]` | -1 | non-contig | 2.47x | 2.61x | 2.54x |

**bfloat16**

| Shape | dim | layout | topk k=8 | topk k=256 | kthvalue k=N/2 |
|---|---|---|---|---|---|
| `[8, 256]` | -1 | contiguous | 1.36x | 1.41x | 1.41x |
| `[8, 1024]` | -1 | contiguous | 1.47x | 1.44x | 1.44x |
| `[8, 4096]` | -1 | contiguous | 1.69x | 1.62x | 1.68x |
| `[8, 16384]` | -1 | contiguous | 1.89x | 1.86x | 1.89x |
| `[8, 65536]` | -1 | contiguous | 2.19x | 2.20x | 2.16x |
| `[128, 256]` | -1 | contiguous | 5.76x | 5.84x | 5.56x |
| `[128, 1024]` | -1 | contiguous | 2.81x | 2.85x | 2.73x |
| `[128, 4096]` | -1 | contiguous | 2.03x | 2.03x | 2.00x |
| `[128, 16384]` | -1 | contiguous | 1.42x | 1.41x | 1.41x |
| `[128, 65536]` | -1 | contiguous | 2.87x | 2.84x | 2.89x |
| `[1024, 256]` | -1 | contiguous | 7.63x | 7.66x | 7.56x |
| `[1024, 1024]` | -1 | contiguous | 2.61x | 2.61x | 2.57x |
| `[1024, 4096]` | -1 | contiguous | 2.06x | 2.07x | 2.00x |
| `[1024, 16384]` | -1 | contiguous | 1.07x | 1.02x | 1.09x |
| `[1024, 65536]` | -1 | contiguous | 1.55x | 1.54x | 1.63x |
| `[8, 256]` | -1 | non-contig | 1.34x | 1.36x | 1.78x |
| `[8, 1024]` | -1 | non-contig | 1.43x | 1.47x | 1.60x |
| `[8, 4096]` | -1 | non-contig | 1.47x | 1.58x | 1.18x |
| `[8, 16384]` | -1 | non-contig | 1.67x | 1.73x | 1.53x |
| `[8, 65536]` | -1 | non-contig | 2.08x | 2.06x | 1.95x |
| `[128, 256]` | -1 | non-contig | 5.79x | 5.60x | 5.74x |
| `[128, 1024]` | -1 | non-contig | 2.85x | 2.84x | 2.61x |
| `[128, 4096]` | -1 | non-contig | 1.87x | 1.83x | 1.75x |
| `[128, 16384]` | -1 | non-contig | 1.33x | 1.33x | 1.30x |
| `[128, 65536]` | -1 | non-contig | 2.65x | 2.64x | 2.34x |
| `[1024, 256]` | -1 | non-contig | 7.67x | 7.73x | 7.30x |
| `[1024, 1024]` | -1 | non-contig | 2.64x | 2.60x | 2.50x |
| `[1024, 4096]` | -1 | non-contig | 1.89x | 1.90x | 1.87x |
| `[1024, 16384]` | -1 | non-contig | 0.79x | 0.98x | 0.85x |
| `[1024, 65536]` | -1 | non-contig | 1.48x | 1.40x | 1.73x |

**int32**

| Shape | dim | layout | topk k=8 | topk k=256 | kthvalue k=N/2 |
|---|---|---|---|---|---|
| `[8, 256]` | -1 | contiguous | 6.70x | 6.54x | 5.95x |
| `[8, 1024]` | -1 | contiguous | 5.58x | 5.65x | 5.89x |
| `[8, 4096]` | -1 | contiguous | 2.61x | 2.59x | 2.51x |
| `[8, 16384]` | -1 | contiguous | 1.61x | 1.63x | 1.42x |
| `[8, 65536]` | -1 | contiguous | 1.27x | 1.27x | 1.16x |
| `[128, 256]` | -1 | contiguous | 19.77x | 17.16x | 17.85x |
| `[128, 1024]` | -1 | contiguous | 9.45x | 9.48x | 8.87x |
| `[128, 4096]` | -1 | contiguous | 2.52x | 2.52x | 2.55x |
| `[128, 16384]` | -1 | contiguous | 0.96x | 0.95x | 0.93x |
| `[128, 65536]` | -1 | contiguous | 1.73x | 1.67x | 1.59x |
| `[1024, 256]` | -1 | contiguous | 35.19x | 34.18x | 31.61x |
| `[1024, 1024]` | -1 | contiguous | 10.14x | 10.09x | 9.96x |
| `[1024, 4096]` | -1 | contiguous | 2.60x | 2.61x | 2.59x |
| `[1024, 16384]` | -1 | contiguous | 0.95x | 0.96x | 0.98x |
| `[1024, 65536]` | -1 | contiguous | 1.90x | 1.84x | 1.74x |
| `[8, 256]` | -1 | non-contig | 6.61x | 5.96x | 6.62x |
| `[8, 1024]` | -1 | non-contig | 5.70x | 5.64x | 5.75x |
| `[8, 4096]` | -1 | non-contig | 2.70x | 2.78x | 2.40x |
| `[8, 16384]` | -1 | non-contig | 1.63x | 1.64x | 1.60x |
| `[8, 65536]` | -1 | non-contig | 1.44x | 1.44x | 1.28x |
| `[128, 256]` | -1 | non-contig | 19.57x | 18.39x | 18.75x |
| `[128, 1024]` | -1 | non-contig | 9.49x | 9.53x | 8.30x |
| `[128, 4096]` | -1 | non-contig | 2.56x | 2.53x | 2.43x |
| `[128, 16384]` | -1 | non-contig | 1.09x | 1.09x | 1.03x |
| `[128, 65536]` | -1 | non-contig | 2.10x | 2.24x | 2.03x |
| `[1024, 256]` | -1 | non-contig | 35.59x | 33.34x | 36.41x |
| `[1024, 1024]` | -1 | non-contig | 10.32x | 10.33x | 9.07x |
| `[1024, 4096]` | -1 | non-contig | 2.67x | 2.65x | 2.40x |
| `[1024, 16384]` | -1 | non-contig | 1.05x | 1.06x | 1.04x |
| `[1024, 65536]` | -1 | non-contig | 2.37x | 2.44x | 2.62x |

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Sort.mm [hunk 1]
```diff
@@ -296,7 +316,8 @@ static void sort_multi_block(const Tensor& input,
   // mb_sort_block has stable variants; mb_merge doesn't (stable on stable input).
   const auto block_fn = fmt::format("mb_sort_block_{}_tptg{}{}{}", type_str, tptg, u16_sfx, stable_sfx);
   const auto merge_fn = fmt::format("mb_merge_{}_tptg{}{}", type_str, tptg, u16_sfx);
-  const auto final_fn = fmt::format("mb_merge_final_{}_tptg{}{}", type_str, tptg, u16_sfx);
+  const char* final_kth_sfx = kth_mode ? "_kth" : "";
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Sort.mm]: Could technically be a std::string_view I guess?

File: aten/src/ATen/native/mps/operations/Sort.mm [hunk 2]
```diff
@@ -457,12 +405,13 @@ void kthvalue_out_mps_impl(const Tensor& self, int64_t k, int64_t dim, Tensor& v
     return;
   }
 
+  const bool kth_mode = kth_target >= 0;
   Tensor out_vals = values;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Sort.mm]: Why inc ref here if we are going to discard right after if we need_copy_back?

File: test/test_mps.py [hunk 3]
```diff
@@ -6287,6 +6287,17 @@ def helper(n, c):
 
         helper(3, 3)
 
+    def _check_kthvalue_matches_cpu(self, cpu, dim):
+        mps = cpu.to("mps")
+        n = cpu.size(dim)
+        for k in (1, max(1, n // 2), n):
```
Comment (Skylion007) [inline test/test_mps.py]: Use parameterize over k so you get proper testcases

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@claude review this in an adversarial fasion, specifically tell me if merge and kthvalue kernels can be somehow merged together. And whether it makes sense to keep kth_kernel rather than just have a topk?
